torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // scriptCall is only alive within this block, use reference to avoid copy
       auto& stack = scriptCall.stackRef();
+
       if (scriptCall.hasOp()) {
         scriptCall.op()->getOperation()(stack);
-      } else {
-        PythonRpcHandler::getInstance()
-            .jitCompilationUnit()
-            ->get_function(scriptCall.qualifiedName())
-            .run(stack);
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
+            stack.size());
+        return wrap(
+            std::move(ScriptResp(std::move(stack.front()))).toMessage());
       }
 
-      TORCH_INTERNAL_ASSERT(
-          stack.size() == 1,
-          "Return value of a builtin operator or a "
-          "TorchScript function should be a single IValue, got a vector of "
-          "size ",
-          stack.size());
-
-      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
+      if (future->completed()) {
+        return wrap(std::move(ScriptResp(future->value())).toMessage());
```

Comment:
What is the benefit of separating this case out? Can we consolidate this into `addCallback`? Is the purpose to avoid an additional func invocation?

Reply:
I expect that probably the majority of the time (i.e. it's just computation, no fork/wait), future will already be completed.

It's more efficient to just instantiate a completed future than to go through the motions of making  an unfilled future, setting up the callback, completing the future, and making the followup callbacks.

Reply:
ah, got it, thanks!